### ISSUE TITLE
Annex B RegExp character class decimal escape in TypeScript 2.1.4

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1275]
Normally I don't report real-world regexp breakages anymore because I know that nonstandard regexp support is still a work in progress.  However, since this breaks the TypeScript compiler it seemed worthwhile to call it out.  The offending regexp is:
```js
var tripleSlashDirectiveFragmentRegex = /^(\/\/\/\s*<reference\s+(path|types)\s*=\s*(?:'|"))([^\3"]*)$/;
```

I don't know if there are more broken regexps, but that's the one that caused an error for me.  It was a bit too convoluted for me to try to find the offending token :)

**Comment by svaarala** [comment]
Ok, I'll try to figure it out.

**Comment by svaarala** [comment]
I think it's the `[^\3"]` part at the very end.

A class ranges may only contain a `\\` as part of a ClassEscape, and the only form allowing digit(s) to follow is a ClassEscape -> DecimalEscape, which only allows `\0` and rejects other decimal escapes:
- http://www.ecma-international.org/ecma-262/6.0/#sec-classescape: Evaluate DecimalEscape, SyntaxError if result is not a character
- http://www.ecma-international.org/ecma-262/6.0/#sec-decimalescape: For number 0 return NUL character, otherwise return integer `i` (which gets rejected).

I'm not 100% sure how Node.js handles it. At the outset it seems that the `\3` is treated like U+0003:

```
/(.)(.)(.)[^\3]/
> x.test('xyzx');
true
> x.test('xyzz');  // not treated like backref; \3 = 'z'
true
> x.test('xyz\u0002');
true
> x.test('xyz\u0003');
false
> x.test('xyz\u0004');
true
```

If this is the case, the fix is to use:

```
[^\3"]  -->  [^\u0003"]
```

**Comment by svaarala** [comment]
If the above interpretation is correct, it should be relatively easy to allow `\NNN` to be parsed as decimal and used as a decimal codepoint number.

On closer inspection it's maybe interpreted as an octal number and treated as a character literal if it's not octal:

```
> x = /[^\9]/
/[^\9]/
> x.test('\u0008')
true
> x.test('\u0009')
true
> x.test('8')
true
> x.test('9')  // matches here
false
```

Still in octal range:
```
> x = /[^\7]/
/[^\7]/
> x.test('\u0007')  // matches here
false
> x.test('\u0008')
true
> x.test('7')  // not here
true
> x.test('8')
true
```

Multiple digits are allowed:

```
> x = /[^\77]/
/[^\77]/
> x.test('\u003f')
false
> x.test('\u003e')
true
```

But it seems to be limited to the usual `0377` range and anything above gets treated as a literal:

```
> x = /[^\777]/
/[^\777]/
> x.test('\u01ff')
true
> x.test('777')  // matches here
false
```

And for anything above, a maximal prefix is matched, e.g. here for `\400` the `\40` gets treated as U+0020, which is then followed by a literal '0':

```
> x = /[^\400]/    // above \0377
/[^\400]/
> x.test('\u0100')   // doesn't match U+0100 == octal 400
true
> x.test('\u0101')
true
> x.test('4')  // not literal '4'
true
> x.test('\u0004')  // nor \4
true
> x.test('\u0020')  // matches here: \40
false
> x.test('1')
true
> x.test('0')  // 0 also matches so after \40 the remaining zero is a literal
false
```

**Comment by svaarala** [comment]
I don't have time now, but I'd like to find the backing from ES6 Annex B for this before changing behavior.

**Comment by svaarala** [comment]
Btw, maybe there's actually a bug in the original RegExp, `\3` in the class ranges would evaluate to U+0003 i.e. ASCII ETX. Or maybe I just misread the whole regexp :-)

**Comment by fatcerberus** [comment]
I doubt very much that the intent was to find U+0003, the regexp seems to be for matching docstrings (i.e. `///` comments).

**Comment by fatcerberus** [comment]
RegEx101 breaks down the regexp as follows:

```js
/^(\/\/\/\s*<reference\s+(path|types)\s*=\s*(?:'|"))([^\3"]*)$/
```
```
^ asserts position at start of the string
  1st Capturing Group (\/\/\/\s*<reference\s+(path|types)\s*=\s*(?:'|"))
    \/ matches the character / literally (case sensitive)
    \/ matches the character / literally (case sensitive)
    \/ matches the character / literally (case sensitive)
    \s* matches any whitespace character (equal to [\r\n\t\f\v ])
    * Quantifier — Matches between zero and unlimited times, as many times as possible, giving back as needed (greedy)
    <reference matches the characters <reference literally (case sensitive)
    \s+ matches any whitespace character (equal to [\r\n\t\f\v ])
    + Quantifier — Matches between one and unlimited times, as many times as possible, giving back as needed (greedy)
    2nd Capturing Group (path|types)
    \s* matches any whitespace character (equal to [\r\n\t\f\v ])
    = matches the character = literally (case sensitive)
    \s* matches any whitespace character (equal to [\r\n\t\f\v ])
    Non-capturing group (?:'|")
  3rd Capturing Group ([^\3"]*)
    Match a single character not present in the list below [^\3"]*
      * Quantifier — Matches between zero and unlimited times, as many times as possible, giving back as needed (greedy)
      \3 matches the character  with index 38 (310 or 316) literally (case sensitive)
      " matches the character " literally (case sensitive)
$ asserts position at the end of the string, or before the line terminator right at the end of the string (if any)
```

**Comment by fatcerberus** [comment]
Errata: That should have been "3<sub>8</sub> (3<sub>10</sub> or 3<sub>16</sub>)" for the `\3` step.  The subscripting was lost in the copy/paste.

**Comment by svaarala** [comment]
Ok, that seems to agree with what I was able to figure out.

**Comment by fatcerberus** [comment]
It very well could be a bug in the the regexp, though I can't imagine what the `\3` was meant to be.

**Comment by fatcerberus** [comment]
Interesting.  From ES2015 Section 21.2.2.9:
> An escape sequence of the form \ followed by a nonzero decimal number n matches the result of the nth set of capturing parentheses (see 0). It is an error if the regular expression has fewer than n capturing parentheses. If the regular expression has n or more capturing parentheses but the nth one is undefined because it has not captured anything, then the backreference always succeeds.

So it's actually a backreference and not a character escape.

**Comment by fatcerberus** [comment]
Also found this, it's apparently a common technique:
http://www.dustindiaz.com/regular-expression-back-matching

**Comment by fatcerberus** [comment]
ES5.1 has the same text re: decimal escapes (i.e. it's a backreference), so it looks like this is actually a bug in *Duktape*, not the regexp. ;-)

**Comment by svaarala** [comment]
Backreferences are not allowed in character classes, and they don't make immediate sense: would you interpret each character in the capture separately?

**Comment by svaarala** [comment]
If you check the links here: https://github.com/svaarala/duktape/issues/1275#issuecomment-271150965, do you read the requirement the same way, i.e. reject a decimal escape other than `\0` in a character class?

**Comment by fatcerberus** [comment]
Ah, you're right, I missed that it was part of a character class.  ES2015 does say quite explicitly that no backreferences are allowed in that context:
> A ClassAtom can use any of the escape sequences that are allowed in the rest of the regular expression except for \b, \B, and backreferences.

Hm.  I'm confused then.  It might help solve this mystery if I had an example of a string that this regex matched.

**Comment by svaarala** [comment]
I think my tests with Node.js indicate pretty conclusively that it is parsed as a legacy octal sequence. It's basically an extension of `\0` which *is* allowed in a character class to denote a NUL character. It then allows you to do specify other ASCII characters using legacy escapes.

**Comment by fatcerberus** [comment]
That still doesn't answer the question of why they'd want to match ASCII ETX though.  Probably a bug like you originally said, maybe I should bring it up on the TypeScript repo.

**Comment by svaarala** [comment]
Oh that, I guess it might be possible that it was thought to be a backreference - but the 3rd capture group in the regexp is the same group where the character class is. So I'm not sure what the goal there was.

**Comment by fatcerberus** [comment]
The comment above that regexp is:

```js
        /**
         * Matches a triple slash reference directive with an incomplete string literal for its path. Used
         * to determine if the caret is currently within the string literal and capture the literal fragment
         * for completions.
         * For example, this matches
         *
         * /// <reference path="fragment
         *
         * but not
         *
         * /// <reference path="fragment"
         */
```

**Comment by fatcerberus** [comment]
Looks like you're right about the legacy octal escape.  Annex B.1.4.1:
> The production CharacterEscape :: LegacyOctalEscapeSequence evaluates by evaluating the SV of the LegacyOctalEscapeSequence (see B.1.2) and returning its character result.

**Comment by fatcerberus** [comment]
Still no idea why they want to match Ctrl+C characters, but the regexp is indeed valid under Annex B semantics it seems like.

**Comment by fatcerberus** [comment]
If I remember correctly, TypeScript 2.0 worked fine under Duktape.  So this must be related to a new feature in v2.1.

**Comment by fatcerberus** [comment]
When I get home I'll check if there are any other regexps in the TypeScript source that break under Duktape.

**Comment by fatcerberus** [comment]
I changed `\3` in the offending regexp to `\u0003` and now TypeScript works, so that seems to be the only issue (for now).

**Comment by svaarala** [comment]
That's good to hear, adding the legacy octal parsing should be relatively straightforward. I'll try to work up a pull tonight.

**Comment by fatcerberus** [comment]
@svaarala Is there still time to get this fix into 2.1.0?  I noticed there's no pull yet.  It's not a showstopper for me, it's just annoying to have to keep editing the TypeScript compiler every time I upgrade it. ;-)

**Comment by svaarala** [comment]
I can try tonight - should be time if it's a straightforward change.
